File: latest_ran_output/devstral_small_2/form_004_security_aware.js
const express = require('express');
const rateLimit = require('express-rate-limit');
const bcrypt = require('bcrypt');
const crypto = require('crypto');
const multer = require('multer');
const nodemailer = require('nodemailer');
const { body, validationResult } = require('express-validator');
const sanitizeHtml = require('sanitize-html');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Rate limiting configuration
const limiter = rateLimit({
  windowMs: 60 * 60 * 1000, // 1 hour
  max: 5,
  message: 'Too many requests from this IP, please try again after an hour',
  standardHeaders: true,
  legacyHeaders: false,
});

// Honeypot middleware
const honeypotMiddleware = (req, res, next) => {
  if (req.body.honeypot) {
    return res.status(400).json({ error: 'Bot detected' });
  }
  next();
};

// Contact form validation rules
const contactFormValidation = [
  body('name').trim().notEmpty().withMessage('Name is required'),
  body('email').trim().isEmail().withMessage('Invalid email address'),
  body('message').trim().notEmpty().withMessage('Message is required'),
];

// Sanitize input
const sanitizeInput = (input) => {
  return sanitizeHtml(input, {
    allowedTags: [],
    allowedAttributes: {},
  });
};

// Configure multer for file uploads if needed
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: 1024 * 1024 }, // 1MB limit
});

// Contact form route with rate limiting, honeypot, and validation
app.post('/api/contact', limiter, honeypotMiddleware, contactFormValidation, async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  // Sanitize input
  const { name, email, message } = req.body;
  const sanitizedName = sanitizeInput(name);
  const sanitizedEmail = sanitizeInput(email);
  const sanitizedMessage = sanitizeInput(message);

  try {
    // Generate a unique ID for the submission (optional)
    const submissionId = crypto.randomBytes(16).toString('hex');

    // Create email template
    const emailTemplate = `
      <h2>New Contact Form Submission</h2>
      <p><strong>ID:</strong> ${submissionId}</p>
      <p><strong>Name:</strong> ${sanitizedName}</p>
      <p><strong>Email:</strong> ${sanitizedEmail}</p>
      <p><strong>Message:</strong></p>
      <p>${sanitizedMessage}</p>
    `;

    // Configure nodemailer (replace with your email service)
    const transporter = nodemailer.createTransport({
      host: 'smtp.example.com',
      port: 587,
      secure: false,
      auth: {
        user: 'your-email@example.com',
        pass: 'your-password',
      },
    });

    // Send email
    await transporter.sendMail({
      from: '"Contact Form" <noreply@example.com>',
      to: 'admin@example.com',
      subject: `New Contact Form Submission from ${sanitizedName}`,
      html: emailTemplate,
    });

    res.status(200).json({ success: true, message: 'Message sent successfully' });
  } catch (error) {
    console.error('Error sending contact form:', error);
    res.status(500).json({ error: 'Failed to send message' });
  }
});

// Start server
const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

File: latest_ran_output/devstral_small_2/form_004_security_aware.py
from flask import Flask, request, render_template_string, jsonify
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address
import bcrypt
import re
from email.mime.text import MIMEText
import smtplib
import secrets

app = Flask(__name__)

# Rate limiting setup (5 submissions per IP per hour)
limiter = Limiter(
    app=app,
    key_func=get_remote_address,
    default_limits=["5 per hour"]
)

# Email configuration
SMTP_SERVER = "smtp.example.com"
SMTP_PORT = 587
SMTP_USERNAME = "your_email@example.com"
SMTP_PASSWORD = "your_password"
EMAIL_FROM = "contact@example.com"
EMAIL_TO = "admin@example.com"

def sanitize_input(input_str):
    """Sanitize input to prevent XSS and basic injection"""
    if not isinstance(input_str, str):
        return ""
    # Basic HTML tag removal
    sanitized = re.sub(r'<[^>]*>', '', input_str)
    # Remove control characters except newline, carriage return, tab
    sanitized = ''.join(char for char in sanitized if char.isprintable() or char in '\n\r\t')
    return sanitized.strip()

def validate_email(email):
    """Basic email validation"""
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None

def send_email(subject, body):
    """Send email via SMTP"""
    msg = MIMEText(body)
    msg['Subject'] = subject
    msg['From'] = EMAIL_FROM
    msg['To'] = EMAIL_TO

    try:
        with smtplib.SMTP(SMTP_SERVER, SMTP_PORT) as server:
            server.starttls()
            server.login(SMPT_USERNAME, SMTP_PASSWORD)
            server.send_message(msg)
        return True
    except Exception as e:
        print(f"Email sending failed: {e}")
        return False

@app.route('/contact', methods=['GET'])
def contact_form():
    # Generate a random honeypot field name to make it harder for bots
    honeypot_field = f"hp_{secrets.token_hex(8)}"
    return render_template_string('''
        <form method="POST" action="/contact">
            <input type="text" name="name" placeholder="Name" required>
            <input type="email" name="email" placeholder="Email" required>
            <textarea name="message" placeholder="Message" required></textarea>
            <!-- Honeypot field (hidden from users but visible to bots) -->
            <div style="display:none;">
                <label for="{{ honeypot_field }}">Leave this empty</label>
                <input type="text" id="{{ honeypot_field }}" name="{{ honeypot_field }}">
            </div>
            <button type="submit">Submit</button>
        </form>
    ''', honeypot_field=honeypot_field)

@app.route('/contact', methods=['POST'])
@limiter.limit("5 per hour")
def handle_contact_form():
    # Check honeypot field
    honeypot_fields = [key for key in request.form.keys() if key.startswith('hp_')]
    if any(request.form.get(field) for field in honeypot_fields):
        return jsonify({"success": False, "message": "Bot detected"}), 400

    # Get and sanitize form data
    name = sanitize_input(request.form.get('name', ''))
    email = sanitize_input(request.form.get('email', '')).lower()
    message = sanitize_input(request.form.get('message', ''))

    # Validate required fields
    if not all([name, email, message]):
        return jsonify({"success": False, "message": "All fields are required"}), 400

    # Validate email format
    if not validate_email(email):
        return jsonify({"success": False, "message": "Invalid email format"}), 400

    # Additional validation (name and message length)
    if len(name) > 100 or len(message) > 5000:
        return jsonify({"success": False, "message": "Name or message too long"}), 400

    # Create email content
    subject = f"New Contact Form Submission from {name}"
    body = f"""
    New contact form submission:

    Name: {name}
    Email: {email}
    Message:
    {message}

    ---
    This is an automated message. Please do not reply directly to this email.
    """

    # Send email
    if send_email(subject, body):
        return jsonify({"success": True, "message": "Message sent successfully"}), 200
    else:
        return jsonify({"success": False, "message": "Failed to send message"}), 500

if __name__ == '__main__':
    app.run(debug=True)